packages/sdk/src/protocols/store.ts

```diff
@@ -0,0 +1,317 @@
+  }
+
+  /**
+   * Do a query to a Waku Store to retrieve historical/missed messages.
+   * The callback function takes a `Promise<WakuMessage>` in input,
+   * useful if messages need to be decrypted and performance matters.
+   *
+   * The order of the messages passed to the callback is as follows:
+   * - within a page, messages are expected to be ordered from oldest to most recent
+   * - pages direction depends on { @link QueryOptions.pageDirection }
+   *
+   * Do note that the resolution of the `Promise<WakuMessage | undefined` may
+   * break the order as it may rely on the browser decryption API, which in turn,
+   * may have a different speed depending on the type of decryption.
+   *
+   * @throws If not able to reach a Waku Store peer to query,
+   * or if an error is encountered when processing the reply,
+   * or if two decoders with the same content topic are passed.
+   */
+  async queryWithPromiseCallback<T extends IDecodedMessage>(
+    decoders: IDecoder<T>[],
+    callback: (
+      message: Promise<T | undefined>
+    ) => Promise<void | boolean> | boolean | void,
+    options?: waku_store.QueryOptions
+  ): Promise<void> {
+    let abort = false;
+    for await (const page of this.queryGenerator(decoders, options)) {
+      const _promises = page.map(async (msgPromise) => {
+        if (abort) return;
+        abort = Boolean(await callback(msgPromise));
+      });
+
+      await Promise.all(_promises);
+      if (abort) break;
+    }
+  }
+
+  createCursor(message: IDecodedMessage): Cursor {
+    if (
+      !message ||
+      !message.timestamp ||
+      !message.payload ||
+      !message.contentTopic
+    ) {
+      throw new Error("Message is missing required fields");
+    }
+
+    const contentTopicBytes = utf8ToBytes(message.contentTopic);
+
+    const digest = sha256(concat([contentTopicBytes, message.payload]));
+
+    const messageTime = BigInt(message.timestamp.getTime()) * BigInt(1000000);
+
+    return {
+      digest,
+      pubsubTopic: message.pubsubTopic,
+      senderTime: messageTime,
+      receiverTime: messageTime
+    };
+  }
+
+  private validateDecodersAndPubsubTopic<T extends IDecodedMessage>(
+    decoders: IDecoder<T>[],
+    options?: waku_store.QueryOptions
+  ): {
+    pubsubTopic: string;
+    contentTopics: string[];
+    decodersAsMap: Map<string, IDecoder<T>>;
+  } {
+    if (decoders.length === 0) {
+      throw new Error("No decoders provided");
+    }
+
+    // convert array to set to remove duplicates
+    const uniquePubsubTopicsInQuery = Array.from(
+      new Set(decoders.map((decoder) => decoder.pubsubTopic))
+    );
+    // If multiple pubsub topics are provided, throw an error
+    if (uniquePubsubTopicsInQuery.length > 1) {
+      throw new Error(
+        "API does not support querying multiple pubsub topics at once"
+      );
+    }
+
+    // we can be certain that there is only one pubsub topic in the query
+    const pubsubTopicForQuery = uniquePubsubTopicsInQuery[0];
```

Comment:
how certain are we? I think maybe some precaution makes sense

Reply:
good point. maybe there's a cleaner way to address this, but this is prefaced with that justifies the last comment:
```ts
  // convert array to set to remove duplicates
    const uniquePubsubTopicsInQuery = Array.from(
      new Set(decoders.map((decoder) => decoder.pubsubTopic))
    );
    // If multiple pubsub topics are provided, throw an error
    if (uniquePubsubTopicsInQuery.length > 1) {
      throw new Error(
        "API does not support querying multiple pubsub topics at once"
      );
    }

    // we can be certain that there is only one pubsub topic in the query
    const pubsubTopicForQuery = uniquePubsubTopicsInQuery[0];
```